### ISSUE TITLE
Source API

### ISSUE THREAD

**Issue by chrismatix** [airyhq/airy#2274]
**Is your feature request related to a problem? Please describe.**
As part of the custom sources milestone (see for motivation), we need an app that exposes a common interface for all custom source connectors. 

**Describe the solution you'd like**
A Kafka streaming component that exposes the following functionality:

- [x] Manage source token permissions
- [x] Create sources
- [x] Ingest messages and metadata
- [x] Create channels
- [x] React to outbound messages

**Additional context**

All interactions with this app need to be authenticated so that you can run custom source apps from outside the cluster, too. For this purpose "creating a source" means: 
- Provision a JWT that encodes the source identifier and does not expire so that the source app can use it as an API key for interactions
- Create the necessary webhook (to be done by the source app)

**Comment by steffh** [comment]
@chrismatix - Thank you for the detailed and super qualified response! Just briefly in return: 

## 2. Removing / updating a Source?

- Great!

## 3. Creating a Channel

- Thank you & understood! I like your compromise `sources.channels.create`. Also fully understood your argument about auth and I think it makes sense and is pretty clever.

- The argument that we use camelCase in URI schemes six times starting since April is inconsistent with us using snake-case in all other places in the API e.g. request bodies, etc. 
- Furthermore using `camelCase` is violating the standard definitions for URI schemes according to RFC 3986 Section 3.1: https://datatracker.ietf.org/doc/html/rfc3986 - According to RFC3986, URLs are “case sensitive” (except for the scheme and the host). In practice, though, a sensitive case may create dysfunctions with APIs e.g. hosted on a Windows system. Therefore, it is generally recommended to use the spinal-case in URIs with hyphens (which is highlighted by RFC3986), this case is used by Google, PayPal, and other big companies.

## 4. Make "Creating a Channel" optional

Okay, then let's carve this out for the time being and let's create a separate ticket ay you suggested. 

## 5. Ingesting messaging data

> In other places *_at timestamps are represented as ISO8601, which is less portable for clients than a unix timestamp. So there's a consistency trade-off:
> 
> Option A: Call it sent_at, but it's a different data type/format
> Option B: Call it sent_at_millis in one place and sent_at in another

Understood, thank you for explaining the difference here, I just think the naming is a bit confusing to people if we mix ISO8601 for `xxx_at` and unix timestamps for `xxx_at_millis`.

As we are talking about timestamps here, could we simply use `timestamp` as the name of the property? Compare with Facebook Message Event Payload: https://developers.facebook.com/docs/messenger-platform/reference/webhook-events/messages/

In the context of ingestion what would be actually great is if we could support both types here, as some sources like `facebook` send Unix timestamps and others RFC3339 UTC such as `google` (https://developers.google.com/business-communications/business-messages/reference/rest/v1/Message).

So could we support both options via the Source API in parallel removing the requirement to convert the timestamps to the other formats? So one of `send_at` (ISO8601) and `timestamp` (UNIX timestamp) needs to be defined. Or to expensive?

**Comment by chrismatix** [comment]
@steffh cool, we're narrowing it down 🎉 

> The argument that we use camelCase in URI schemes six times starting since April is inconsistent with us using snake-case in all other places in the API e.g. request bodies, etc.. Furthermore using camelCase is violating the standard definitions for URI schemes according to RFC 3986 Section 3.1: https://datatracker.ietf.org/doc/html/rfc3986 - According to RFC3986, URLs are “case sensitive” (except for the scheme and the host). In practice, though, a sensitive case may create dysfunctions with APIs e.g. hosted on a Windows system. Therefore, it is generally recommended to use the spinal-case in URIs with hyphens (which is highlighted by RFC3986), this case is used by Google, PayPal, and other big companies.

It's not in violation of `3.1`, because it does not relate to the scheme. But I'm fine with the convention argument so I will also create a ticket for changing the other URLs to hyphenation. 

> So could we support both options via the Source API in parallel removing the requirement to convert the timestamps to the other formats? So one of send_at (ISO8601) and timestamp (UNIX timestamp) needs to be defined. Or too expensive?

I'd prefer to call it `send_at` then (to keep it consistent with the existing data model) and make it work with both `ISO8601` and a unix timestamp. For now, I will go for those two formats, since `RFC3986` is a bit out there.

**Comment by steffh** [comment]
@chrismatix - Perfect!

**Comment by steffh** [comment]
@chrismatix - Just a question: What was the rationale to rename `POST /sources.ingest` in the draft version to `POST /sources.webhook`?

I think "ingest" was more descriptive than "webhook" and less confusing - as we already use "webhooks" as a destination to consume data (https://airy.co/docs/core/api/webhook).

Maybe I also still lack some insight into the runtime you wanted to create which I assume would also have an external "/sourcename/webhook" route, then remap that data and call the `POST /sources.ingest` route from within the runtime.

**Comment by chrismatix** [comment]
@steffh Great question! When I tried assuming the role of an outside developer `ingest` started to look cryptic to me, because `(Data) ingestion` is a very abstract term that can denote many ways of getting data into a system. Whereas `webhook` is an exact description of what this endpoint is.

So in this case I prefer `clarity > disambiguation`.

But we can add the following line to the docs to make it extra clear to users:
```
Note: This webhook is for data inbound to the Airy system. To register an outbound webhook have a look at the Webhook API (link).
```

Curious to hear what you think.
